File: src/main/java/com/music/music_inventory_api/service/GenreService.java
package com.music.music_inventory_api.service;

import com.music.music_inventory_api.dto.request.CreateGenreRequest;
import com.music.music_inventory_api.dto.request.UpdateGenreRequest;
import com.music.music_inventory_api.dto.response.AlbumResponse;
import com.music.music_inventory_api.dto.response.GenreResponse;
import java.util.List;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for Genre entity operations. Defines business logic for
 * managing genres.
 */
public interface GenreService {

    /**
     * Create a new genre.
     *
     * @param request
     *                the genre creation request
     * @return the created genre response
     */
    GenreResponse createGenre(CreateGenreRequest request);

    /**
     * Get genre by ID.
     *
     * @param id
     *           the genre ID
     * @return the genre response
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *                                                                         if
     *                                                                         genre
     *                                                                         not
     *                                                                         found
     */
    GenreResponse getGenreById(Long id);

    /**
     * Get all genres with pagination.
     *
     * @param pageable
     *                 pagination information
     * @return page of genre responses
     */
    Page<GenreResponse> getAllGenres(Pageable pageable);

    /**
     * Update an existing genre.
     *
     * @param id
     *                the genre ID
     * @param request
     *                the genre update request
     * @return the updated genre response
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *                                                                         if
     *                                                                         genre
     *                                                                         not
     *                                                                         found
     */
    GenreResponse updateGenre(Long id, UpdateGenreRequest request);

    /**
     * Delete a genre by ID.
     *
     * @param id
     *           the genre ID
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *                                                                         if
     *                                                                         genre
     *                                                                         not
     *                                                                         found
     */
    void deleteGenre(Long id);

    /**
     * Get all albums by genre ID.
     *
     * @param genreId
     *                the genre ID
     * @return list of album responses
     * @throws com.music.music_inventory_api.exception.EntityNotFoundException
     *                                                                         if
     *                                                                         genre
     *                                                                         not
     *                                                                         found
     */
    List<AlbumResponse> getAlbumsByGenre(Long genreId);
}

File: frontend/src/services/genreService.ts
import api from './api';

export interface Genre {
    id: number;
    name: string;
    description?: string;
}

export interface CreateGenreRequest {
    name: string;
    description?: string;
}

export interface UpdateGenreRequest {
    name?: string;
    description?: string;
}

class GenreService {
    private readonly endpoint = '/genres';

    async getAll(): Promise<Genre[]> {
        const response = await api.get<Genre[]>(this.endpoint);
        return response.data;
    }

    async getById(id: number): Promise<Genre> {
        const response = await api.get<Genre>(`${this.endpoint}/${id}`);
        return response.data;
    }

    async create(genreData: CreateGenreRequest): Promise<Genre> {
        const response = await api.post<Genre>(this.endpoint, genreData);
        return response.data;
    }

    async update(id: number, genreData: UpdateGenreRequest): Promise<Genre> {
        const response = await api.put<Genre>(`${this.endpoint}/${id}`, genreData);
        return response.data;
    }

    async delete(id: number): Promise<void> {
        await api.delete(`${this.endpoint}/${id}`);
    }
}

export default new GenreService();
